lib/lazyrecord.rb

```diff
@@ -137,3 +149,9 @@ def save
     self.class.save(self)
   end
 end
+
+
+set_trace_func proc { |event, _file, _line, _id, _binding, classname|
+  # only interested in events of type 'call' of LazyRecord
+  classname.send('gateway') if event == 'call' && classname == 'LazyRecord'
+}
```

Comment:
This is the first time I see this 🤔 

Isn't there a way to override the send method from the class?

Something like:
```
class LazyRecord
  def self.send
    unless @ready
       gateway
    end
    super
  end
end
```

Reply:
There I'm trying to execute a global 'hook' before any other method to ensure the class is ready. Another alternative could be just to call the gateway method inside another method call but that would be too much coupling.

The other solution is to use modules like ActiveRecord and execute[ callbacks ](https://github.com/rails/rails/blob/v3.2.21/activerecord/lib/active_record/callbacks.rb).

What do you think @julionav ?